File: backend/app/com/skewwhiffy/auraltester/scales/ScaleTypeFactory.scala
package com.skewwhiffy.auraltester.scales

import com.skewwhiffy.auraltester.internalnotation.IntervalFactory
import com.skewwhiffy.auraltester.notes.DirectedInterval

import javax.inject.{Inject, Singleton}

@Singleton
class ScaleTypeFactory @Inject()(intervalFactory: IntervalFactory) {
  private lazy val majorIntervals: String = "1 2 3 4 5 6 7 8"
  private lazy val minorHarmonicIntervals: String = "1 2 3- 4 5 6- 7 8"
  private lazy val minorMelodicAscendingIntervals: String = "1 2 3- 4 5 6 7 8"
  private lazy val minorMelodicDescendingIntervals: String = "1 2 3- 4 5 6- 7- 8"

  def major: ScaleType = new ScaleType("major", "", majorIntervals.toIntervals)
  def minorHarmonic: ScaleType = new ScaleType("minor harmonic", "min", minorHarmonicIntervals.toIntervals)
  def minorMelodic: ScaleType = new ScaleType(
    "minor melodic",
    "min",
    minorMelodicAscendingIntervals.toIntervals,
    minorMelodicDescendingIntervals.toIntervals
  )

  implicit class IntervalConverter(val rawIntervals: String) {
    def toIntervals: List[DirectedInterval] = intervalFactory.getDirectedIntervals(rawIntervals)
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/notation/factory/ScaleTypeFactory.kt
package com.skewwhiffy.notation.factory

import com.skewwhiffy.notation.model.scale.ScaleType
import org.springframework.stereotype.Service

@Service
class ScaleTypeFactory(private val intervalFactory: IntervalFactory) {
  private val majorIntervals: String = "1 2 3 4 5 6 7 8"
  private val minorHarmonicIntervals: String = "1 2 3- 4 5 6- 7 8"
  private val minorMelodicAscendingIntervals: String = "1 2 3- 4 5 6 7 8"
  private val minorMelodicDescendingIntervals: String = "1 2 3- 4 5 6- 7- 8"

  val major get() = ScaleType("major", "", majorIntervals.toIntervals)

  val minorHarmonic get() = ScaleType("minor harmonic", "min", minorHarmonicIntervals.toIntervals)

  val minorMelodic
    get() = ScaleType(
      "minor melodic",
      "min",
      minorMelodicAscendingIntervals.toIntervals,
      minorMelodicDescendingIntervals.toIntervals
    )

  private val String.toIntervals get() = intervalFactory.getDirectedIntervals(this)

}